### PR TITLE
Add hypertrace instrumentation

### PR DESCRIPTION
**Pull request by freeall**
# Hypercore speedtest

All these tests have been performed by running `$ npm test` in hypercore on my macbook M1

In each case, it was run 10 times and an average was counted.

## Summary

There was no significant difference between running on an instrumented (but not actively tracing) code base vs a non-instrumented one.

When starting to trace (i.e. `setTraceFunction()` have been called) there was a ~40% increase in overhead.

When starting to trace, but using cache, by `.trace(cacheId)` there was a ~10% increase in overhead.

In real-life cases with long-lived objects the overhead is insignificant.

## 1. Without hypertrace instrumentation added

This is from the `main` hypercore branch without any instrumentation added

```
# time = 10845.309917ms
# time = 10951.91625ms
# time = 10950.779916ms
# time = 11360.088333ms
# time = 9467.811375ms
# time = 9663.833167ms
# time = 10812.367917ms
# time = 11270.643792ms
# time = 9504.236916ms
# time = 9521.476208ms
# ======================
# avg  = 10434.8463791ms
```

### 2a. Without using setTraceFunction

Using `createTracer()`, `.trace()`, but not `setTraceFunction()`

```
# time = 10157.106042ms
# time = 11162.466041ms
# time = 10852.541959ms
# time = 10037.762541ms
# time = 11161.77325ms
# time = 11033.951375ms
# time = 10692.505ms
# time = 11217.457667ms
# time = 11003.730792ms
# time = 9489.278625ms
# ======================
# avg  = 10680.8573292ms
```

### 2b. While using setTraceFunction(() => { })), and also using a cacheId

Using `createTracer()`, `.trace(cacheId)`, and `setTraceFunctio()`.

All calls to `.trace()` have a cache id similar to the function name:
``` js
  send () {
    this.tracer.trace('send')
    ...
  }
```

```
# time = 10983.720042ms
# time = 10053.00875ms
# time = 10984.808667ms
# time = 9447.403042ms
# time = 10707.983666ms
# time = 10521.029334ms
# time = 9876.588375ms
# time = 11113.740125ms
# time = 11134.92525ms
# time = 10559.54175ms
# ===================
# avg  = 10538.2749001ms
```

### 2c. While using setTraceFunction(() => { }), but no cacheId

Using `createTracer()`, `.trace()`, and `setTraceFunction()`.

This is added to the very first line of `index.js`:

``` js
const { setTraceFunction } = require('hypertrace')
let tracesCount = 0
setTraceFunction(() => { tracesCount++ })
process.on('exit', () => console.log(`trace function called ${tracesCount} times`))
```

There was an average of 251,540 calls to the trace function when running the tests

```
# time = 14284.143125ms
# time = 14546.467542ms
# time = 14444.38875ms
# time = 14550.258416ms
# time = 14664.504209ms
# time = 13550.680334ms
# time = 14850.7915ms
# time = 14498.501041ms
# time = 13862.527833ms
# time = 14399.782458ms
# ======================
# avg  = 14365,2045208ms
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,3 +1,7 @@
+const { setTraceFunction } = require('hypertrace')
+let tracesCount = 0
+setTraceFunction(() => { tracesCount++ })
+process.on('exit', () => console.log(`trace function called ${tracesCount} times`))
```
Comment (HDegroote) [inline index.js:4]: Maybe only console.log if `tracesCount > 0`? So we don't have that log unless we're tracing

Comment (mafintosh) [inline index.js:4]: shouldnt be in here obvs

File: index.js [hunk 2]
```diff
@@ -846,6 +852,7 @@ module.exports = class Hypercore extends EventEmitter {
   }
 
   async get (index, opts) {
+    this.tracer.trace()
```
[the remark this answers is not in the document]
Comment (mafintosh) [reply index.js:855]: and lets trace this post-ready, ie after the line below, then the ordering makes more sense

File: index.js [hunk 3]
```diff
@@ -950,6 +957,7 @@ module.exports = class Hypercore extends EventEmitter {
   }
 
   download (range) {
+    this.tracer.trace()
```
Comment (mafintosh) [inline index.js:960]: move this one to _download and trace after opening and pass the range, { range } to props

File: index.js [hunk 4]
```diff
@@ -995,6 +1003,7 @@ module.exports = class Hypercore extends EventEmitter {
   }
 
   async append (blocks, opts = {}) {
+    this.tracer.trace()
```
Comment (mafintosh) [inline index.js:1006]: move here https://github.com/holepunchto/hypercore/pull/465/files#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R1015 and pass the blocks when its guaranteed to be an array

File: lib/replicator.js [hunk 5]
```diff
@@ -267,7 +268,8 @@ class BlockTracker {
 }
 
 class Peer {
-  constructor (replicator, protomux, channel, session) {
+  constructor (replicator, protomux, channel, session, coreTracer) {
+    this.tracer = createTracer(this, { parent: coreTracer })
```
Comment (mafintosh) [inline lib/replicator.js:272]: just do parent: replicator.core.tracer

File: lib/replicator.js [hunk 6]
```diff
@@ -564,6 +568,8 @@ class Peer {
   }
 
   async onrequest (msg) {
+    this.tracer.trace()
```
Comment (mafintosh) [inline lib/replicator.js:571]: pass the msg as props, ie `.tracer.trace(msg)`

File: lib/replicator.js [hunk 7]
```diff
@@ -661,6 +667,8 @@ class Peer {
   }
 
   async ondata (data) {
+    this.tracer.trace()
```
Comment (mafintosh) [inline lib/replicator.js:670]: pass the data as props

File: lib/replicator.js [hunk 8]
```diff
@@ -720,6 +728,8 @@ class Peer {
   }
 
   onnodata ({ request }) {
+    this.tracer.trace()
```
Comment (mafintosh) [inline lib/replicator.js:731]: pass { onrequest } as props

File: lib/replicator.js [hunk 9]
```diff
@@ -1161,6 +1171,8 @@ class Peer {
   }
 
   async _send (req) {
+    this.tracer.trace()
```
Comment (mafintosh) [inline lib/replicator.js:1174]: move this one here https://github.com/holepunchto/hypercore/pull/465/files#diff-3344a84994d8e172768f33ec4bf514ba2bc28eb7eeffff10fcbbd294efe56edcR1204 and pass the req as props

File: lib/replicator.js [hunk 10]
```diff
@@ -2067,7 +2080,7 @@ module.exports = class Replicator {
 
     if (channel === null) return onnochannel()
 
-    const peer = new Peer(replicator, protomux, channel, session)
+    const peer = new Peer(replicator, protomux, channel, session, this.core.tracer)
```
Comment (mafintosh) [inline lib/replicator.js:2083]: remove the argument and just get it from the replicator arg (see above comment)

File: index.js (hunk 1 above)
Comment (freeall) [reply index.js:4]: Yup

File: lib/replicator.js [hunk 11]
```diff
@@ -564,6 +568,8 @@ class Peer {
   }
 
   async onrequest (msg) {
+    this.tracer.trace({ msg })
```
Comment (mafintosh) [inline lib/replicator.js:571]: just pass msg and not { msg }

File: lib/replicator.js [hunk 12]
```diff
@@ -661,6 +667,8 @@ class Peer {
   }
 
   async ondata (data) {
+    this.tracer.trace({ data })
```
Comment (mafintosh) [inline lib/replicator.js:670]: just pass data

File: lib/replicator.js [hunk 13]
```diff
@@ -1190,6 +1200,8 @@ class Peer {
       return
     }
 
+    this.tracer.trace({ req })
```
Comment (mafintosh) [inline lib/replicator.js:1203]: just pass req

File: index.js [hunk 14]
```diff
@@ -266,6 +268,8 @@ module.exports = class Hypercore extends EventEmitter {
     this.writable = this._isWritable()
     this.autoClose = o.autoClose
 
+    this.tracer.setParent(o.core?.tracer)
```
Comment (mafintosh) [inline index.js:271]: no need for this, the below ones are always ran

File: lib/replicator.js [hunk 15]
```diff
@@ -1190,6 +1200,8 @@ class Peer {
       return
     }
 
+    this.tracer.trace('_send', req)
```
Comment (mafintosh) [inline lib/replicator.js:1203]: just 'send' is fine

**Pull request merged**